File: Python/matrixBuilders/coordinate_matrix_generator.py
import random
from matrix.coomatrix import CoordinateMatrix
from matrix.coordinates import Coordinate

class CoordinateMatrixGenerator():
    
    @staticmethod
    def generate_random_coordinate_matrix(n, density):
        if density < 0.0 or density > 1.0:
            raise ValueError("Density must be in the range [0, 1].")

        total_non_zero_elements = int(round(n * n * density))
        random_instance = random.Random()

        coordinates = set()

        while len(coordinates) < total_non_zero_elements:
            row = random_instance.randint(0, n - 1)
            col = random_instance.randint(0, n - 1)
            coordinates.add((row, col))

        return CoordinateMatrix(n, [Coordinate(row, col, random_instance.randint(1, 100)) for row, col in coordinates])

    @staticmethod
    def contains_coordinate(coordinates, row, col):
        for coordinate in coordinates:
            if coordinate.i == row and coordinate.j == col:
                return True
        return False

File: src/main/java/matrixmultiplication/matrixbuilder/CoordinateMatrixGenerator.java
package matrixmultiplication.matrixbuilder;

import matrixmultiplication.Matrix;
import matrixmultiplication.MatrixBuilder;
import matrixmultiplication.matrix.Coordinate;
import matrixmultiplication.matrix.CoordinateMatrix;

import java.util.*;

public class CoordinateMatrixGenerator implements MatrixBuilder {

    public static CoordinateMatrix generateRandomCoordinateMatrix(int n, double density) {
        if (density < 0.0 || density > 1.0) {
            throw new IllegalArgumentException("Density must be in the range [0, 1].");
        }

        List<Coordinate> coordinates = new ArrayList<>();
        Random random = new Random();
        int totalNonZeroElements = (int) Math.round(n * n * density);
        Set<Integer> usedCoordinates = new HashSet<>();

        while (coordinates.size() < totalNonZeroElements) {
            int row, col;
            do {
                row = random.nextInt(n);
                col = random.nextInt(n);
            } while (!usedCoordinates.add(row * n + col));

            int value = random.nextInt(101);
            coordinates.add(new Coordinate(row, col, value));
        }

        return new CoordinateMatrix(n, coordinates);
    }

    @Override
    public void set(int i, int j, int value) {

    }

    @Override
    public void setMatrix(Matrix c) {
    }

    @Override
    public Matrix get() {
        return null;
    }
}